lib/scholar/naive_bayes/gaussian_naive_bayes.ex

```diff
@@ -0,0 +1,442 @@
+      class_count: final_class_count,
+      class_priors: class_priors,
+      classes: classes,
+      epsilon: eps
+    }
+  end
+
+  defnp update_mean_variance_init(mu, var, x, sample_weights, mask, sample_weights_flag) do
+    {num_samples, _num_features} = Nx.shape(x)
+
+    cond do
+      num_samples == 0 ->
+        {mu, var, 0}
+
+      sample_weights_flag == false ->
+        new_n = num_samples
+        new_mu = mean_masked(x, mask)
+        new_var = mean_masked((x - new_mu) ** 2, mask)
+        {new_mu, new_var, new_n}
+
+      true ->
+        new_n = Nx.sum(sample_weights * mask)
+        new_mu = mean_weighted_masked(x, mask, sample_weights)
+        new_var = mean_weighted_masked((x - new_mu) ** 2, mask, sample_weights)
+        {new_mu, new_var, new_n}
+    end
+  end
+
+  defnp update_mean_variance(past_n, new_n, new_mu, new_var, mu, var) do
+    total_n = past_n + new_n
+    total_mu = (new_n * new_mu + past_n * mu) / total_n
+
+    past_ssd = past_n * var
+    new_ssd = new_n * new_var
+    total_ssd = past_ssd + new_ssd + new_n * past_n / total_n * (mu - new_mu) ** 2
+    total_var = total_ssd / total_n
+
+    {total_mu, total_var}
+  end
+
+  defnp joint_log_likelihood(
+          %__MODULE__{class_priors: class_priors, var: var, theta: theta},
+          x
+        ) do
+    pi = 3.14159265359
+    joint = Nx.log(class_priors)
+    {num_classes, num_features} = Nx.shape(theta)
+    {samples_x, _features_x} = Nx.shape(x)
+
+    n1 =
+      (-0.5 * Nx.sum(Nx.log(2.0 * pi * var), axes: [1]))
+      |> Nx.reshape({:auto, 1})
+      |> Nx.broadcast({num_classes, samples_x})
+
+    x_broadcast =
+      Nx.reshape(x, {1, samples_x, num_features})
+      |> Nx.broadcast({num_classes, samples_x, num_features})
+
+    theta_broadcast =
+      Nx.reshape(theta, {num_classes, 1, num_features})
+      |> Nx.broadcast({num_classes, samples_x, num_features})
+
+    var_broadcast =
+      Nx.reshape(var, {num_classes, 1, num_features})
+      |> Nx.broadcast({num_classes, samples_x, num_features})
+
+    n2 = -0.5 * Nx.sum((x_broadcast - theta_broadcast) ** 2 / var_broadcast, axes: [2])
+
+    (n1 + n2) |> Nx.transpose() |> Nx.add(joint)
+  end
+
+  defnp mean_masked(t, mask) do
+    {num_samples, num_features} = Nx.shape(t)
+
+    broadcast_mask =
+      mask |> Nx.reshape({num_samples, 1}) |> Nx.broadcast({num_samples, num_features})
+
+    Nx.sum(t * broadcast_mask, axes: [0]) / Nx.sum(broadcast_mask, axes: [0])
+  end
+
+  defnp mean_weighted_masked(t, mask, weights) do
+    {num_samples, num_features} = Nx.shape(t)
+
+    broadcast_mask =
+      mask |> Nx.reshape({num_samples, 1}) |> Nx.broadcast({num_samples, num_features})
+
+    broadcast_weights =
+      weights |> Nx.reshape({num_samples, 1}) |> Nx.broadcast({num_samples, num_features})
+
+    Nx.sum(t * broadcast_mask * broadcast_weights, axes: [0]) /
+      Nx.sum(broadcast_mask * broadcast_weights, axes: [0])
+  end
+
+  defnp check_input(%__MODULE__{theta: theta}, x) do
```

Comment:
given this is basically a transform, perhaps deftransformp would be more fitting (no pun intended)
```suggestion
  deftransformp check_input(%__MODULE__{theta: theta}, x) do
```

Reply:
If it works as defnp, i think we should stick with it, because it guarantees certain properties.